subprojects/core/src/main/java/org/gradle/internal/classpath/TypeHierarchyRegistry.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath;
+
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.NonNullApi;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.gradle.internal.classanalysis.AsmConstants.ASM_LEVEL;
+
+@NonNullApi
+public class TypeHierarchyRegistry {
+
+    private final Map<String, Set<String>> superTypes = new ConcurrentHashMap<>();
+    private final Map<String, Set<String>> directSuperTypes = new ConcurrentHashMap<>();
+
+    public Set<String> getSuperTypes(String type) {
+        return Collections.unmodifiableSet(superTypes.computeIfAbsent(type, this::computeSuperTypes));
+    }
+
+    public void visit(byte[] classBytes) {
+        ClassReader classReader = new ClassReader(classBytes);
+        classReader.accept(new TypeHierarchyClassVisitor(this), 0);
+    }
+
+    private void registerSuperType(String className, @Nullable String superType) {
+        if (superType != null) {
+            directSuperTypes.computeIfAbsent(className, k -> ConcurrentHashMap.newKeySet()).add(superType);
+        }
+    }
+
+    private void registerInterfaces(String className, String[] interfaces) {
+        for (String superType : interfaces) {
+            directSuperTypes.computeIfAbsent(className, k -> ConcurrentHashMap.newKeySet()).add(superType);
+        }
+    }
+    private Set<String> computeSuperTypes(String type) {
```

Comment:
I believe walking the graph never occurs concurrently with modifying the `directSuperTypes` map? We can probably emphasize that by adding a flag that is flipped at some point, making the registry effectively immutable.

Reply:
It's probably true, that `TypeHierarchyRegistry` is never accessed concurrently. 

But I think we can't mark it as immutable anywhere, since we core types lazily at the moment, e.g. when `getSuperTypes` is called. 🤔